### PR TITLE
Elements: Headings

### PR DESCRIPTION
**Pull request by clrux**
# Headings
This work provides base pattern library styling for headings (`H1` - `H6`) along with specific cases for the headings (e.g., `emphasized` or `deemphasized`). I've taken care to match the output to the mockup by any means necessary.

## What's included?
* Pixels for `font-size` and `line-height` variables. These are converted to `rems` in the function. Doing this allows us to match the mockup exactly. I was having issues getting to be precise using the `modular-scale` function from Bourbon.
* I've adjusted the colors of the headings (specifically the `gray` variables) to pass AA accessibility guidelines.
* I'm currently including two different methods of deemphasizing, but am not sold on either. Thoughts here are welcome. Both classes are included right now with the intention of removing or replacing after some discussion.
* I added a `letter-spacing` map and function, but I don't think I decided to use it. When I was futzing with the `modular-scale` stuff, the letters looked too spaced out. We can keep it or remove it.

---

@frrrances I believe this is ready for your review. Let me know if you have questions or if you need help getting the Gulp setup working.
cc @talbs

### CODE REVIEW

File: _src/static/sass/components/_headings.scss [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+// ----------------------------
+.hd-6 {
+    font-size: font-size(base);
+    line-height: line-height(base);
+
+    &.deemphasized,
+    &.secondary {
+        color: palette(grayscale, mid-dark);
+    }
+}
```
Comment (clrux) [inline _src/static/sass/components/_headings.scss]: I included both `deemphasized` and `secondary` here, unsure of which I like better, if either of them. The purpose is to downplay whatever heading class they've chosen. So, while our normal headings might be slightly bolder, these options will make them less bold, or lighter gray, etc.

File: _src/static/sass/utilities/_variables.scss [hunk 2]
```diff
@@ -193,34 +193,31 @@ $font-family-monospace: 'Bitstream Vera Sans Mono', Consolas, Courier, monospace
...
-    large:          modular-scale(2, $font-scale-base),
-    x-large:        modular-scale(3, $font-scale-base),
-    xx-large:       modular-scale(4, $font-scale-base)
+    xx-large:       43,
+    x-large:        32,
+    large:          24,
+    mid-large:      21,
+    mid-small:      18,
+    base:           $font-scale-base
 );
```
Comment (clrux) [inline _src/static/sass/utilities/_variables.scss]: I was having a really tough time with modular-scale. I couldn't get them to precisely match up to the mockups. I decided to get more control by supplying a pixel size (matching the mockup) and then lettering Bourbon convert it to `rems`. The end result is the same, especially if the mockup was designed using Modular Scale.

File: _src/static/sass/utilities/_variables.scss [hunk 3]
```diff
@@ -193,34 +193,31 @@ $font-family-monospace: 'Bitstream Vera Sans Mono', Consolas, Courier, monospace
...
-    x-large:        1.48
+    xx-large:       58,
+    x-large:        47,
+    large:          36,
+    mid-large:      31,
+    mid-small:      26,
+    small:          24,
+    base:           24,
+    none:           0
 );
```
Comment (clrux) [inline _src/static/sass/utilities/_variables.scss]: See my comment above with Modular Scale as it's applicable here.

File: _src/static/sass/utilities/_variables.scss [hunk 4]
```diff
@@ -233,6 +230,17 @@ $font-weights: (
...
+$letter-spacing: (
+    normal:         0,
+    tightest:       -3,
+    tighter:        -2,
+    tight:          -.095rem,
+    loose:          -.075rem,
+    looser:         -.025rem,
+    loosest:        3
+);
+
```
Comment (clrux) [inline _src/static/sass/utilities/_variables.scss]: Aware of the... naming. It's unused at this point. If we decide to keep it we can re-think the map variables.

File: _posts/elements/2015-02-04-example-kitchen-sink.md [hunk 5]
```diff
@@ -17,33 +17,33 @@ info:  Vestibulum id ligula porta felis euismod semper. Donec id elit non mi por
```
Comment (talbs) [inline _posts/elements/2015-02-04-example-kitchen-sink.md]: I think this post can be deleted - it was originally created to make sure my PL doc site styling was scoped to not affect PL elements.

Comment (talbs) [inline _posts/elements/2015-02-04-example-kitchen-sink.md]: We can delete it in a separate PR - thanks for trying to update it.

File: _posts/elements/2015-03-25-example-headings-new.md [hunk 6]
```diff
@@ -0,0 +1,50 @@
+---
```
Comment (talbs) [inline _posts/elements/2015-03-25-example-headings-new.md]: Can you change the title of this post to just be "headings"?

Comment (talbs) [inline _posts/elements/2015-03-25-example-headings-new.md]: Can you change the name of this file to be ``2015-03-25-headings.md``?

File: _posts/elements/2015-03-25-example-headings-new.md [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+layout:       element
+title:        Example - Demo Headings
+date:         2015-02-25 00:00:00
+
+categories:   elements
+tags:
+- example
```
Comment (talbs) [inline _posts/elements/2015-03-25-example-headings-new.md]: Please remove the example tag (it was made for the placeholder posts I added when getting the side rendering up and running).

Comment (talbs) [inline _posts/elements/2015-03-25-example-headings-new.md]: Please add the atomic tag, since headings are an atomic pattern.

File: _posts/elements/2015-03-25-example-headings-new.md [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+slug:         example-headings-new
```
Comment (talbs) [inline _posts/elements/2015-03-25-example-headings-new.md]: Can you change the slug of this post to just be "headings"?

File: _posts/elements/2015-03-25-example-headings-new.md [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+scssurl:      "https://github.com/edx/ux-pattern-library/blob/master/[path_to_source_sass_file].scss"
+
+description:  Examples of headings in a semi-real world context.
+
+info:  Vestibulum id ligula porta felis euismod semper. Donec id elit non mi porta gravida at eget metus. Integer posuere erat a ante venenatis dapibus posuere velit aliquet. Etiam porta sem malesuada magna mollis euismod.
+---
+
+<section class="group">
```
Comment (talbs) [inline _posts/elements/2015-03-25-example-headings-new.md]: I'd prefer us to keep the markup as lean (just the elements and any supporting elements/classes/attributes as needed) as possible for these examples. Can you remove the ``<section>`` and ``<article>`` elements here?

File: _posts/elements/2015-03-25-example-headings-new.md [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+    <article>
+        <h1 class="hd-1">Proident laborum ut nulla consequat kevin adipisicing dolar.</h1>
+        <p>In dolor filet mignon est. Flank tail eu, pork belly sed incididunt shank short ribs id pig sausage meatball. Ribeye landjaeger pancetta pork belly shank officia incididunt in dolor excepteur swine ham. In ut short ribs consectetur hamburger ipsum.</p>
```
Comment (talbs) [inline _posts/elements/2015-03-25-example-headings-new.md]: Please reduce this to just one paragraph of text (its good to see how a headings is rendered against copy, but I don't think we need so much).

Comment (talbs) [inline _posts/elements/2015-03-25-example-headings-new.md]: Please give the ``<p>`` element a class of copy (i.e. ``<p class="copy">``) for now (so that it will pick up any base copy styling).

Comment (talbs) [inline _posts/elements/2015-03-25-example-headings-new.md]: Actually, never mind that business, a plain ``<p>`` is fine. I'll delete this thread.

File: _src/static/sass/_build.scss [hunk 11]
```diff
@@ -21,7 +21,7 @@
 // ------------------------------
 // #COMPONENTS
 // ------------------------------
-// TBD
+@import 'components/headings';
```
Comment (talbs) [inline _src/static/sass/_build.scss]: Terra incognita no more! Yay!

File: _src/static/sass/components/_headings.scss [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+
+// About: Contains base styling for headings, relying on Bourbon's mathematical scales for precision.
+
+// Scale:
+// 12, 14, 16, 18, 21, 24, 32, 43
+// ----------------------------
+
+// #GLOBALS
+// #HEADINGS
+// H1
```
Comment (talbs) [inline _src/static/sass/components/_headings.scss]: I don't think these H1-6 are needed in the ToC/initial comments. Please remove.

File: _src/static/sass/utilities/_functions.scss [hunk 13]
```diff
@@ -27,7 +27,7 @@
 // get line-heights from defined map values
 @function line-height($key) {
   @if map-has-key($line-heights, $key) {
-    @return map-get($line-heights, $key);
+    @return rem(map-get($line-heights, $key));
```
Comment (talbs) [inline _src/static/sass/utilities/_functions.scss]: Please remove this addition. We're not adding any units of measurements on purpose (see in our styleguide here - https://github.com/edx/ux-pattern-library/wiki/Styleguide:-Sass-&-CSS#line-heights).

File: _src/static/sass/utilities/_functions.scss [hunk 14]
```diff
@@ -44,6 +44,16 @@
   @return null;
 }
 
+// get letter spacing from defined map values
```
Comment (talbs) [inline _src/static/sass/utilities/_functions.scss]: Sweet function, bro!

File: _src/static/sass/base/_typography.scss [hunk 15]
```diff
@@ -48,6 +48,13 @@ p, ol, ul, dl, td, th {
     @extend %reset-copy;
 }
 
+p,
+li,
+dt,
+dd {
+    line-height: line-height(base);
```
Comment (talbs) [inline _src/static/sass/base/_typography.scss]: This should be covered by the rule above that leverages ``@extend %reset-copy;``. Right now that extend is using ``line-height: line-height(large);``. If you need to change it, I'd change it there.

Comment (talbs) [inline _src/static/sass/base/_typography.scss]: If you're cool with that, please remove this rule as it doesn't seem needed.

File: _src/static/sass/utilities/_variables.scss [hunk 16]
```diff
@@ -193,34 +193,31 @@ $font-family-monospace: 'Bitstream Vera Sans Mono', Consolas, Courier, monospace
...
 
-$font-scale-small: 16;
-$font-scale-medium: 18;
+$font-scale-small: 14;
+$font-scale-medium: 16;
 $font-scale-large: 18;
 $font-scale-base: $font-scale-medium;
 
-// typography: sizes
+// typography: sizes (pixels)
```
Comment (talbs) [inline _src/static/sass/utilities/_variables.scss]: Please remove the "(pixels)" note from the comment here. I don't think its needed as we we are just defining a scale.

File: _src/static/sass/utilities/_variables.scss (hunk 2 above)
Comment (talbs) [inline _src/static/sass/utilities/_variables.scss]: Sounds good to me. Thanks for trying with Modular scale and if its giving us trouble, I'm fine with leaving plain and clear with these values.

File: _src/static/sass/utilities/_variables.scss (hunk 3 above)
Comment (talbs) [inline _src/static/sass/utilities/_variables.scss]: Sounds good to me. Thanks for trying with Modular scale and if its giving us trouble, I'm fine with leaving plain and clear with these values.

File: _src/static/sass/utilities/_variables.scss (hunk 4 above)
Comment (talbs) [inline _src/static/sass/utilities/_variables.scss]: I'd prefer if we use the same naming convention we used in other scales/maps (see https://github.com/edx/ux-pattern-library/blob/master/_src/static/sass/utilities/_variables.scss#L215).

Comment (talbs) [inline _src/static/sass/utilities/_variables.scss]: Also, please loose the ``rem`` units from this scale, I'd prefer them to be added in the ``letter-spacing`` function.

File: _src/static/sass/components/_headings.scss [hunk 17]
```diff
@@ -0,0 +1,128 @@
...
+
+// ----------------------------
+// #GLOBALS
+// ----------------------------
+.hd-1,
+.hd-2,
+.hd-3,
+.hd-4,
+.hd-5,
+.hd-6 {
```
Comment (talbs) [inline _src/static/sass/components/_headings.scss]: Is there any benefit into making this into a Sass extend or adjusting the ``%reset-headings`` extend (https://github.com/edx/ux-pattern-library/blob/master/_src%2Fstatic%2Fsass%2Futilities%2F_helpers.scss#L95) for this basic rule.

File: _src/static/sass/components/_headings.scss [hunk 18]
```diff
@@ -0,0 +1,128 @@
...
+    font-family: $headings-base-font-family;
+    font-weight: $headings-base-font-weight;
+    color: $headings-base-color;
+
+    &.emphasized,
```
Comment (talbs) [inline _src/static/sass/components/_headings.scss]: Is this a place (with ``.emphasized``) to use stateful classes or not?

File: _src/static/sass/components/_headings.scss [hunk 19]
```diff
@@ -0,0 +1,128 @@
...
+    &.primary {
+        font-weight: $headings-primary-font-weight;
+        color: $headings-primary-color;
+    }
+
+    &.deemphasized,
```
Comment (talbs) [inline _src/static/sass/components/_headings.scss]: Is this a place (with ``.deemphasized``) to use stateful classes or not?

Comment (talbs) [inline _src/static/sass/components/_headings.scss]: NIT: rename ``deemphasized`` to ``de-emphasized``.

File: _src/static/sass/components/_headings.scss [hunk 20]
```diff
@@ -0,0 +1,128 @@
...
+
+
+// ----------------------------
+// H6
+// ----------------------------
+.hd-6 {
+    font-size: font-size(base);
+    line-height: line-height(base);
+
+    &.deemphasized,
```
Comment (talbs) [inline _src/static/sass/components/_headings.scss]: See comments above about ``.deemphasized``.

File: _src/static/sass/components/_headings.scss [hunk 21]
```diff
@@ -0,0 +1,128 @@
...
+
+
+// ----------------------------
+// H5
+// ----------------------------
+.hd-5 {
+    font-size: font-size(mid-small);
+    line-height: line-height(mid-small);
+
+    &.deemphasized,
```
Comment (talbs) [inline _src/static/sass/components/_headings.scss]: See comments above about ``.deemphasized``.

File: _src/static/sass/components/_headings.scss [hunk 22]
```diff
@@ -0,0 +1,128 @@
...
+
+// ----------------------------
+// H4
+// ----------------------------
+.hd-4 {
+    font-weight: font-weight(bold);
+    font-size: font-size(mid-large);
+    line-height: line-height(mid-large);
+
+    &.deemphasized,
```
Comment (talbs) [inline _src/static/sass/components/_headings.scss]: See comments above about ``.deemphasized``.

File: _src/static/sass/components/_headings.scss [hunk 23]
```diff
@@ -0,0 +1,128 @@
...
+
+
+// ----------------------------
+// H3
+// ----------------------------
+.hd-3 {
+    font-size: font-size(large);
+    line-height: line-height(large);
+
+    &.deemphasized,
```
Comment (talbs) [inline _src/static/sass/components/_headings.scss]: See comments above about ``.deemphasized``.

File: _src/static/sass/components/_headings.scss (hunk 1 above)
Comment (talbs) [inline _src/static/sass/components/_headings.scss]: Good question. I tend to like the simplicity of a two level scale for headings:

* base - default headings (including more bolder weights/colors for the more higher level headings)
* de-emphasized - for the more subdued/downplayed versions

Primary and secondary seem create a more flexible but not as clear (especially with our class names for headings also adding a # scale, i.e. .hd-1-6) scale. I'd recommend losing the primary and secondary classes.

File: _src/static/sass/components/_headings.scss (hunk 18 above)
Comment (talbs) [inline _src/static/sass/components/_headings.scss]: Also, revisiting this, is this class/rule needed? I thought we were creating a base set of headings and then just allowing a a de-emphasized version of those (i.e. no emphasized third version). If this is the case, please remove.

File: _src/static/sass/utilities/_variables.scss (hunk 4 above)
Comment (clrux) [inline _src/static/sass/utilities/_variables.scss]: Loose rems! Annnd they're gone.

File: _src/static/sass/components/_headings.scss [hunk 24]
```diff
@@ -0,0 +1,128 @@
...
+.hd-6 {
+    font-weight: $headings-primary-font-weight;
+    text-transform: uppercase;
+}
+
+
+// ----------------------------
+// #HEADINGS
+// ----------------------------
+// H1
```
Comment (talbs) [inline _src/static/sass/components/_headings.scss]: NIT: these comments shouldn't have this formatting. They should match that of https://github.com/edx/ux-pattern-library/blob/master/_src%2Fstatic%2Fsass%2Futilities%2F_variables.scss#L283. So more like this:

````scss
// ----------------------------
// #HEADINGS
// ----------------------------
// Level one heading
.hd-1 {
    ...
}

// Level two heading
.hd-2 {
    ...
}

...

// Level six heading
.hd-6 {
    ...
}

````

Comment (talbs) [inline _src/static/sass/components/_headings.scss]: Also, as shown in the above example, please rename the headings to avoid any confusion between these classes and semantic heading tags (I want to avoid the "Do H2 styles only go on ``<h2>`` elements?" type of questions).

File: _src/static/sass/utilities/_variables.scss [hunk 25]
```diff
@@ -96,14 +96,14 @@ $palettes: (
     grayscale:           (
         base:       $gray,
         mid-light:  tint($gray, 40%),
-        light:      tint($gray, 60%),
-        x-light:    tint($gray, 80%),
+        light:      tint($gray, 50%),
```
Comment (talbs) [inline _src/static/sass/utilities/_variables.scss]: Why were these changed? If they're to address a11y concerns, cool. If for some other reason, what was it?

File: _src/static/sass/utilities/_variables.scss [hunk 26]
```diff
@@ -96,14 +96,14 @@ $palettes: (
...
+        x-light:    tint($gray, 75%),
         xx-light:   tint($gray, 90%),
         white:      tint($gray, 98%),
         white-t:    tint($gray, 100%),
         mid-dark:   shade($gray, 40%),
-        dark:       shade($gray, 60%),
-        x-dark:     shade($gray, 80%),
+        dark:       shade($gray, 50%),
```
Comment (talbs) [inline _src/static/sass/utilities/_variables.scss]: Why were these changed? If they're to address a11y concerns, cool. If for some other reason, what was it?

File: _src/static/sass/utilities/_variables.scss (hunk 25 above)
Comment (talbs) [inline _src/static/sass/utilities/_variables.scss]: Nevermind, this is a11y related. Thanks for caring and syncing.

File: _src/static/sass/utilities/_variables.scss (hunk 26 above)
Comment (talbs) [inline _src/static/sass/utilities/_variables.scss]: Nevermind, this is a11y related. Thanks for caring and syncing.

File: _src/static/sass/utilities/_variables.scss [hunk 27]
```diff
@@ -193,34 +193,31 @@ $font-family-monospace: 'Bitstream Vera Sans Mono', Consolas, Courier, monospace
...
-    mid-large:      modular-scale(1, $font-scale-base),
-    large:          modular-scale(2, $font-scale-base),
-    x-large:        modular-scale(3, $font-scale-base),
-    xx-large:       modular-scale(4, $font-scale-base)
+    xx-large:       43,
+    x-large:        32,
+    large:          24,
+    mid-large:      21,
+    mid-small:      18,
+    base:           $font-scale-base
```
Comment (talbs) [inline _src/static/sass/utilities/_variables.scss]: Thinking here - having the ``$font-scale-base`` variable defined and used here is less valuable now that we're not using to to populate the rest of the scale. The biggest win was that in our media queries we could change that variable to a smaller value and get a whole new scale for smaller viewports. This isn't the case now, so we may want to remove it in the future.

**Comment by talbs** [comment]
@clrux, thanks for the good write up/set up. I've looked over the code side of things and left my thoughts. I'll be reviewing the visual rendering now and will note any things I find on that level.

Let me know if you have any questions and when you want me to spin through again.

File: _src/static/sass/utilities/_variables.scss (hunk 27 above)
Comment (talbs) [inline _src/static/sass/utilities/_variables.scss]: @clrux, it looks like the font scale here is off from what was designed in https://openedx.atlassian.net/browse/UX-1524. You may want to check the Sketch file (https://drive.google.com/a/edx.org/folderview?id=0B7jcFrBngjmkTGdpTnNJb29faUE&usp=sharing) and inspect each heading's text properties. 

I'll try to add notes inline where things are off in this map.

File: _src/static/sass/utilities/_variables.scss [hunk 28]
```diff
@@ -193,34 +193,31 @@ $font-family-monospace: 'Bitstream Vera Sans Mono', Consolas, Courier, monospace
...
-    base:           $font-scale-base,
-    mid-small:      modular-scale(-1, $font-scale-base),
-    small:          modular-scale(-2, $font-scale-base),
-    x-small:        modular-scale(-3, $font-scale-base),
-    xx-small:       modular-scale(-4, $font-scale-base),
-    mid-large:      modular-scale(1, $font-scale-base),
-    large:          modular-scale(2, $font-scale-base),
-    x-large:        modular-scale(3, $font-scale-base),
-    xx-large:       modular-scale(4, $font-scale-base)
+    xx-large:       43,
```
Comment (talbs) [inline _src/static/sass/utilities/_variables.scss]: xx-large should be 38.

Comment (talbs) [inline _src/static/sass/utilities/_variables.scss]: x-large should be 28

File: _src/static/sass/utilities/_variables.scss (hunk 27 above)
Comment (talbs) [inline _src/static/sass/utilities/_variables.scss]: base should be 18. This also needs to be changed in https://github.com/edx/ux-pattern-library/blob/clrux/elements-headings/_src%2Fstatic%2Fsass%2Futilities%2F_variables.scss#L18.

File: _src/static/sass/utilities/_variables.scss (hunk 2 above)
Comment (talbs) [inline _src/static/sass/utilities/_variables.scss]: ``small`` should be brought back and assigned to 14

File: _src/static/sass/utilities/_variables.scss [hunk 29]
```diff
@@ -193,34 +193,31 @@ $font-family-monospace: 'Bitstream Vera Sans Mono', Consolas, Courier, monospace
...
+    x-large:        32,
+    large:          24,
+    mid-large:      21,
+    mid-small:      18,
```
Comment (talbs) [inline _src/static/sass/utilities/_variables.scss]: This value doesn't match the scale any more - ``mid-small`` should be smaller than base (which is currently 16). Please address.

**Comment by talbs** [comment]
@clrux, one other note, since the scale I designed is based off of 18 (see the sketch file and the output screens), you'll have to update the ``$em-base`` value (here - https://github.com/edx/ux-pattern-library/blob/clrux/elements-headings/_src%2Fstatic%2Fsass%2Futilities%2F_variables.scss#L18) to be 18.

File: _src/static/sass/components/_headings.scss (hunk 17 above)
Comment (clrux) [inline _src/static/sass/components/_headings.scss]: I've moved this into the reset extend. In fact, the extend had most of this so it was just removed.

**Comment by clrux** [comment]
@talbs Ready for another review from you, sir.

**Comment by talbs** [comment]
@clrux, also, it looks like you deleted both of the compiled CSS files we were previously checking in - public/css/main-ltr.css and public/css/main-rtl.css. Can you add those back?

**Comment by talbs** [comment]
@clrux, here's a quick PR to remove the example elements and clean up bits - https://github.com/edx/ux-pattern-library/pull/20. Merge when you want or not.

**Comment by talbs** [comment]
@clrux, here's another PR to add in the 3 levels of heading - base, emphasis, and de-emphasis - https://github.com/edx/ux-pattern-library/pull/21

**Comment by talbs** [comment]
@clrux, one visual question - did you add the proper margins to each of these headings? If not, you can find those details in the visual design's sketch file and in this static comp - https://openedx.atlassian.net/secure/attachment/16459/headings--spacing%2Bvisual-dividers.png.

**Comment by clrux** [comment]
Margins have been in addressed in 88d199f.

File: _src/static/sass/components/_headings.scss [hunk 30]
```diff
@@ -0,0 +1,99 @@
...
+    line-height: line-height(x-large);
+    margin: 0 0 spacing-vertical(small) 0;
+}
+
+
+// Level three heading
+.hd-3 {
+    font-size: font-size(large);
+    line-height: line-height(large);
+    margin: 0 0 spacing-vertical(x-small) 0;
```
Comment (talbs) [inline _src/static/sass/components/_headings.scss]: NIT: property order.

File: _src/static/sass/components/_headings.scss [hunk 31]
```diff
@@ -0,0 +1,99 @@
...
+    line-height: line-height(xx-large);
+    margin: 0 0 spacing-vertical(small) 0;
+}
+
+
+// Level two heading
+.hd-2 {
+    font-size: font-size(x-large);
+    line-height: line-height(x-large);
+    margin: 0 0 spacing-vertical(small) 0;
```
Comment (talbs) [inline _src/static/sass/components/_headings.scss]: NIT: property order.

File: _src/static/sass/components/_headings.scss [hunk 32]
```diff
@@ -0,0 +1,99 @@
...
+
+
+// ----------------------------
+// #INDIVIDUAL HEADINGS
+// ----------------------------
+// Level one heading
+.hd-1 {
+    font-size: font-size(xx-large);
+    line-height: line-height(xx-large);
+    margin: 0 0 spacing-vertical(small) 0;
```
Comment (talbs) [inline _src/static/sass/components/_headings.scss]: NIT: property order.

File: _src/static/sass/components/_headings.scss [hunk 33]
```diff
@@ -0,0 +1,99 @@
...
+    }
+}
+
+
+// Level four heading
+.hd-4 {
+    font-weight: font-weight(bold);
+    font-size: font-size(mid-large);
+    line-height: line-height(mid-large);
+    margin: 0 0 spacing-vertical(x-small) 0;
```
Comment (talbs) [inline _src/static/sass/components/_headings.scss]: NIT: property order.

File: _src/static/sass/components/_headings.scss [hunk 34]
```diff
@@ -0,0 +1,99 @@
...
+}
+
+
+// Level five heading
+.hd-5 {
+    font-size: font-size(mid-small);
+    line-height: line-height(mid-small);
+    margin: 0 0 spacing-vertical(x-small) 0;
```
Comment (talbs) [inline _src/static/sass/components/_headings.scss]: NIT: property order.

File: _src/static/sass/components/_headings.scss [hunk 35]
```diff
@@ -0,0 +1,99 @@
...
+}
+
+
+// Level six heading
+.hd-6 {
+    font-size: font-size(small);
+    line-height: line-height(small);
+    margin: 0 0 spacing-vertical(x-small) 0;
```
Comment (talbs) [inline _src/static/sass/components/_headings.scss]: NIT: property order.

File: _src/static/sass/components/_headings.scss [hunk 36]
```diff
@@ -0,0 +1,99 @@
...
+// ----------------------------
+// #GLOBAL
+// ----------------------------
+.hd-1,
+.hd-2,
+.hd-3,
+.hd-4,
+.hd-5,
+.hd-6 {
+    @extend %reset-headings;
```
Comment (talbs) [inline _src/static/sass/components/_headings.scss]: Should we reset margins at this level or at the extends level rather than the ``margin: x x Y x`` shorthand syntax? If so, for particular headings we can just use the ``margin-bottom: Y`` property.

File: _src/static/sass/components/_headings.scss (hunk 31 above)
Comment (talbs) [inline _src/static/sass/components/_headings.scss]: See my note about ``margin`` shorthand vs. ``margin-bottom`` earlier.

File: _src/static/sass/components/_headings.scss (hunk 32 above)
Comment (talbs) [inline _src/static/sass/components/_headings.scss]: See my note about ``margin`` shorthand vs. ``margin-bottom`` earlier.

File: _src/static/sass/components/_headings.scss (hunk 30 above)
Comment (talbs) [inline _src/static/sass/components/_headings.scss]: See my note about ``margin`` shorthand vs. ``margin-bottom`` earlier.

File: _src/static/sass/components/_headings.scss (hunk 33 above)
Comment (talbs) [inline _src/static/sass/components/_headings.scss]: See my note about ``margin`` shorthand vs. ``margin-bottom`` earlier.

File: _src/static/sass/components/_headings.scss (hunk 34 above)
Comment (talbs) [inline _src/static/sass/components/_headings.scss]: See my note about ``margin`` shorthand vs. ``margin-bottom`` earlier.

File: _src/static/sass/components/_headings.scss (hunk 35 above)
Comment (talbs) [inline _src/static/sass/components/_headings.scss]: See my note about ``margin`` shorthand vs. ``margin-bottom`` earlier.

**Comment by talbs** [comment]
@clrux, a few more comments/details. Also, remember to change ``$baseline`` to not include the ``rem()`` mixin, please.

**Comment by talbs** [comment]
:+1: once you squash commits and revise the commit message to give a summary of adding headings.

Thanks for all of the attention to detail.
